### ISSUE TITLE
Changing an image on a read-only repository uses wrong image path

### ISSUE THREAD

**Issue by RubenVerborgh** [mavoweb/mavo#235]
Upon uploading a new image on the [editable homepage example](http://mavo.io/demos/homepage/), it replaces the image URL by <code><strong>http://data.mavo.io/images/</strong>myimage.jpg</code> instead of the GitHub URL to where this image was uploaded.

Also, even though this images resulted in a 404, it attempts to download it multiple times (up to 13, it seems) with different timestamps, separated by 2-second waiting intervals.

**Comment by LeaVerou** [comment]
> Upon uploading a new image on the [editable homepage example](http://mavo.io/demos/homepage/), it replaces the image URL by `**http://data.mavo.io/images/**myimage.jpg` instead of the GitHub URL to where this image was uploaded.

It doesn’t always use `data.mavo.io`, it just uses the URL from the original repo, where it should use the fork. It's tough though, cause that means that if the fork gets deleted, the image stops existing, whereas at least now, if the PR gets merged, the image reference will work...

> Also, even though this images resulted in a 404, it attempts to download it multiple times (up to 13, it seems) with different timestamps, separated by 2-second waiting intervals.

Yeah, it tries multiple times because often websites are deployed with commit hooks and the like and changes can take a while to be reflected. It's debatable however whether trying every 2 seconds for 10 times max linearly is a good algorithm. Perhaps a better algorithm would be one where the interval is increasing, for two reasons:
- A 1 second latency is much more perceivable from 1 to 2 seconds than from 5 to 6.
- I'd wager that the probability of an image loading within in the next N seconds is not linear and decreases over time. I.e. it's much more likely that an image will load in the next 2 seconds from the moment of upload than if we've already waited 10 seconds.
@karger, thoughts?

**Comment by LeaVerou** [comment]
(for David who mainly reads mail so doesn't get edits)

- With an increasing interval we can get better loading times for images that deploy fast and give better chances of loading at all for images that deploy slowly, since the total time will increase.

**Comment by RubenVerborgh** [comment]
> It doesn’t always use `data.mavo.io`, it just uses the URL from the original repo

Sure, I didn't mean otherwise—just provided exactly what happened in my example case above.

> if the fork gets deleted, the image stops existing, whereas at least now, if the PR gets merged, the image reference will work

I've just tried this, and it seems that deleting a fork on GitHub automatically closes the associated pull request (but doesn't necessarily prevent still merging it).

In any case, using the fork URL seems most logical; for instance, if I upload an image with the same name, I might not see the update. (Also, if I try that at the moment, the wrong URL is generated: http://data.mavo.io/images/photo.jpg instead of http://mavo.io/demos/homepage/images/photo.jpg, but that might be something different altogether.)

**Comment by LeaVerou** [comment]
> In any case, using the fork URL seems most logical; for instance, if I upload an image with the same name, I might not see the update. 

Note that I'm not saying there's no bug here, I’m just saying it's unclear to me how this can be solved because it's trickier than it seems. Certainly, using the fork URL is **better**, but still creates problems. Uploading images has two components: the actual image file and the URL to it that is put in the data. If the URL is based on the fork, it works for you but will stop working on the upstream repo if you delete your fork. The actual file will still be there, because it will be part of the pull request, but the URL will be pointing to the file in the fork.

> (Also, if I try that at the moment, the wrong URL is generated: http://data.mavo.io/images/photo.jpg instead of http://mavo.io/demos/homepage/images/photo.jpg, but that might be something different altogether.)

That is the correct URL (this bug notwithstanding; i.e. if you weren't in a fork). Uploads are relative to the data file, not the HTML document. The reason that happens is that we certainly know that we can write to a repo where the data file resides. That doesn't necessarily mean we can write to the repo that the HTML file resides, as they may not be the same.

**Comment by RubenVerborgh** [comment]
> but will stop working on the upstream repo if you delete your fork.

Seems acceptable to me: novice users will not know how to delete a fork, and advanced users will understand the consequences.

> That is the correct URL (this bug notwithstanding; i.e. if you weren't in a fork).

Yes, makes sense (I missed the fact that the original `/demos/homepage/images/photo.jpg` was not part of the data repository).

**Comment by LeaVerou** [comment]
> Seems acceptable to me: novice users will not know how to delete a fork, and advanced users will understand the consequences.

In most cases, a fork is an artifact to make it possible to suggest changes. Often more knowledgeable people "clean up" after their PRs are merged by deleting their forks. It's very inconsistent with pretty much everything to say that upstream will *depend* on a fork to function properly.

**Comment by LeaVerou** [comment]
Also, security. The whole point of PRs is that you get to control what you accept from users. If I implemented what you're saying, someone could send a PR that adds some reasonable data, then change the image in their fork to a dick pic and basically deface the initial site.

**Comment by RubenVerborgh** [comment]
> Often more knowledgeable people "clean up" after their PRs are merged

But when it's merged, the image simply becomes part of the upstream, so there is no problem?

Let me write out the scenario:

1. I am on http://mavo.io/demos/homepage/ in view mode. The image is `images/photo.jpg`.
0. I switch to edit mode and change the image to `other.jpg`, which is uploaded to my fork.
0. I save and close my browser.
0. My pull request is accepted, meaning that `other.jpg` is merged to the upstream.
0. I delete my fork.
0. I am on http://mavo.io/demos/homepage/ in view mode. The image is `images/other.jpg`, which is now present in the upstream.
0. I switch to edit mode (which, since my work was deleted, creates a new fork from the latest version, which includes `images/other.jpg`).

Are the above steps correct? If so, I don't see a problematic dependency of the upstream and the fork.

> If I implemented what you're saying, someone could send a PR that adds some reasonable data, then change the image in their fork to a dick pic and basically deface the initial site.

No, only I see my own fork as long as the PR is not accepted. And when the PR is accepted, it is the exact same image that the PR contained.

So maybe there's a misunderstanding here: the image in the fork should only be visible to the owner of the fork (before PR acceptance), just like the text of a fork is only visible to the owner of that fork.

**Comment by LeaVerou** [comment]
I’m talking about using the Mavo uploader, not specifying relative URLs on your own. 

Uploaded images get an absolute URL of one of the two forms:
- If the repo they are in has Github Pages enabled, that URL is used
- Otherwise, a rawgit.com URL is used.

The URL is not relative because while **you** know which repo your site is in, Mavo does not. All that Mavo knows is the repo with the data file, because you've linked to that repo via `mv-storage`. Mavo does not know whether the rest of your site is in there or not, or whether that's just a data repo. In fact, in the mavo.io case, most demos data are hosted on a separate repo and not on the same repo that hosts mavo.io.

**Both** of these types of URLs depend on the repo they are from. A fork will most likely NOT have Github Pages enabled and will fall back to the general rawgit URL. That rawgit URL will be part of the data that is merged upstream. That rawgit URL however depends on the fork, and will not point to the image in the upstream repo even after the PR is merged (unless the URL is manually changed). That rawgit URL will stop working once the fork is deleted.

Does it make sense now?

**Comment by RubenVerborgh** [comment]
Actually, while testing some things to write a reply, I bumped into another error: the image is uploaded, but the path is never saved (I wanted to check which path it got).

Here you can see the addition of the file (through Mavo uploader):

- https://github.com/RubenVerborgh/data/commit/5080a99dedffbe42e068a12b0eeb61063c606d17

And here are two subsequent saves, both of which are empty:

- https://github.com/RubenVerborgh/data/commit/fc550af477cf910fe1b4ede4d7486c77857653bf
- https://github.com/RubenVerborgh/data/commit/ecedd638263601d085de3d998bd29187912528cd

Is this because of the 404 that occurs on http://data.mavo.io/images/other.jpg?
Should I file this as another issue?

**Comment by LeaVerou** [comment]
The path is part of the data and needs to be saved with the data (by clicking Save).
Agreed this is a confusing design, but not a bug per se. What I was trying to avoid is saving other changes that have been made to the data as well since the last save.

**Comment by RubenVerborgh** [comment]
> The path is part of the data and needs to be saved with the data (by clicking Save).

Click save two times, but path didn't update. So that _is_ a bug then?

**Comment by LeaVerou** [comment]
Whoa, yes.

**Comment by LeaVerou** [comment]
Did anything else save?

**Comment by RubenVerborgh** [comment]
Yes, all else saves fine. I suspect the 404 on http://data.mavo.io/images/other.jpg prevents the path from updating.

**Comment by LeaVerou** [comment]
It shouldn't. Wait, what exactly do you mean by "the path didn't update"?

**Comment by RubenVerborgh** [comment]
The old image remains in the DOM (presumably since the new one 404s).

Video: https://youtu.be/uJA4dDFXbuE?t=16s
Corresponding commits:
- https://github.com/RubenVerborgh/data/commit/d99756116f02df0be6f0c07eed5bdf34ae7c918d (sav after editing text)
- https://github.com/RubenVerborgh/data/commit/c1a30a0fe9ca50781e06984bd87638c505102a4f (upload image [was the same though])
- https://github.com/RubenVerborgh/data/commit/a9768db1c0e6abcd63f710fcf6932b51bc176448 (save after image upload)

**Comment by LeaVerou** [comment]
Yes, that looks like a bug. The `data.mavo.io/images/...` URL should have been saved in place of `"images/photo.jpg"`.

**Comment by RubenVerborgh** [comment]
Okay, let's first follow up on #241 before we continue the discussion here, because _some_ path needs to be saved before we can discuss the interpretation of that path.

**Comment by RubenVerborgh** [comment]
Thanks! How does this influence merging pull requests of the data, given that a specific rawgit URL is committed (https://github.com/RubenVerborgh/data/commit/4df616a51cb846becf53abdf8d3e324a4a446502)?

**Comment by LeaVerou** [comment]
It doesn’t because it can’t. They will depend on the fork. :(
I figured this was better than people uploading images that they can’t see.
But this is a much bigger open problem that we need to brainstorm about, probably with @karger too.
